docs/src/explanations/optional-groups.md

```diff
@@ -0,0 +1,177 @@
+
+To declare an optional group, extend the `chisel3.group.Declaration` abstract
+class and specify a convention.  To declare a nested optional group, extend the
+`chisel3.group.Declaration` abstract class inside another declaration.
+
+The following example declares four optional groups:
+
+```scala mdoc:silent
+import chisel3.group.{Convention, Declaration}
+
+object A extends Declaration(Convention.Bind) {
+  object B extends Declaration(Convention.Bind) {
+    object C extends Declaration(Convention.Bind)
+  }
+  object D extends Declaration(Convention.Bind)
+}
+```
+
+A _definition_ of an optional group adds optional functionality to a module.
+Each definition must refer to a declaration of an optional group.  Definitions
+must match the nesting of their declarations.
+
+To define an optional group, use the `chisel3.group` apply method referring to a
+previously declared optional group inside a Chisel module.  An optional group
+definition may use any Chisel or Scala value visible to its Scala lexical scope.
+
+The following example defines four optional groups inside module `Foo` and
+declares wires which are connected to values captured from visible lexical
+scope:
+
+```scala mdoc:silent
+import chisel3._
+import chisel3.group
+
+class Foo extends RawModule {
+  val port = IO(Input(Bool()))
+
+  group(A) {
+    val a = WireInit(port)
+    group(A.B) {
+      val b = WireInit(a)
+      group(A.B.C) {
+        val c = WireInit(b)
+      }
+    }
+    group(A.D) {
+      val d = WireInit(port ^ a)
+    }
+  }
+}
+```
+
+## Conventions
+
+Currently, there is only one supported convetion, `Bind`.  This will cause
+optional group definitions to be lowered to Verilog modules that are
+instantiated via the SystemVerilog `bind` mechanism.  The lowering to Verilog of
+nested groups avoids illegal nested usage of `bind`.
+
+More conventions may be supported in the future.
+
+## Examples
+
+### Design Verification Example
+
+Consider a use case where a design or design verification engineer would like to
+add some asserts and debug prints to a module.  The logic necessary for the
+asserts and debug prints requires additional computation.  All of this code
+should not be included in the final Verilog.  The engineer can use three
+optional groups to do this.
+
+There are three optional groups that emerge from this example:
+
+1. A common `Verification` group
+1. An `Assert` group nested under the `Verification` group
+1. A `Debug` group also nested under the `Verification` group
+
+The `Verification` group can be used to store common logic used by both the
+`Assert` and `Debug` groups.  The latter two groups allow for separation of,
+respectively, assertions from prints.
+
+One way to write this in Scala is the following:
+
+```scala mdoc:reset:silent
+import chisel3._
+import chisel3.group.{Convention, Declaration}
+
+// All groups are declared here.  The Assert and Debug groups are nested under
+// the Verification group.
```

Comment:
Not sure you need to change anything here, but for my own curiousity, I assume the nesting-ness of the object matching the expected nesting usage is just for convention/sanity? Or is it important to nest the objects the same as the usages?

Reply:
It's important that the group definition nesting matches the group declaration nesting. The declaration nesting establishes what groups depend on other groups. In this example, the `Assert` group depends on the `Verification` group, the `Assume` group also depends on the `Verification` group, but the `Assert` and `Assume` group do not depend on each other.

This then needs to be enforced so that a group definition cannot access anything that it is not supposed to. For this example, `Assert` and `Assume` can access anything in `Verification`, but `Verification` cannot access anything in `Assert` or `Assume`.

By establishing the nesting and then enforcing it (after https://github.com/chipsalliance/chisel/pull/3562), this is able to use Scala's lexical scoping rules to ensure that anything accessible in a group from Scala's perspective is accessible in a group from FIRRTL's perspective.

Put differently, this is trying to explicitly disallow the following because this changes the dependencies between the groups (`Assert` now depends on `Assume` and it is capturing things it isn't supposed to):

```scala
group(Verification) {
  group(Verification.Assume) {
    val a = Wire(Bool())
    group (Verification.Assert) {
      val b = WireInit(a)
    }
  }
}
```

We could be slightly more relaxed on the Chisel API side and allow a nested group to be used without its parent, e.g., allow code like the following where there is an implicit wrapping of each of both of these (separately) with a `group(Verification)`:

```scala
group(Verification.Assert) { /* ... */ }
group(Verification.Assume) { /* ... */ }
```

This requires a simple `MergeGroups` pass to work well: https://github.com/llvm/circt/issues/6228